z3/src/rec_func_decl.rs

```diff
@@ -0,0 +1,184 @@
+    ///
+    /// ```
+    /// # use z3::{Config, Context, RecFuncDecl, Solver, Sort, Symbol};
+    /// # let cfg = Config::new();
+    /// # let ctx = Context::new(&cfg);
+    /// let f = RecFuncDecl::new(
+    ///     &ctx,
+    ///     "f",
+    ///     &[&Sort::int(&ctx), &Sort::real(&ctx)],
+    ///     &Sort::int(&ctx));
+    /// assert_eq!(f.arity(), 2);
+    /// ```
+    pub fn arity(&self) -> usize {
+        unsafe { Z3_get_arity(self.ctx.z3_ctx, self.z3_func_decl) as usize }
+    }
+
+    /// Create a constant (if `args` has length 0) or function application (otherwise).
+    ///
+    /// Note that `args` should have the types corresponding to the `domain` of the `RecFuncDecl`.
+    pub fn apply(&self, args: &[&ast::Dynamic<'ctx>]) -> ast::Dynamic<'ctx> {
+        assert!(args.iter().all(|s| s.get_ctx().z3_ctx == self.ctx.z3_ctx));
+
+        let args: Vec<_> = args.iter().map(|a| a.get_z3_ast()).collect();
+
+        ast::Dynamic::new(self.ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_mk_app(
+                self.ctx.z3_ctx,
+                self.z3_func_decl,
+                args.len().try_into().unwrap(),
+                args.as_ptr(),
+            )
+        })
+    }
+
+    /// Return the `DeclKind` of this `RecFuncDecl`.
+    pub fn kind(&self) -> DeclKind {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_get_decl_kind(self.ctx.z3_ctx, self.z3_func_decl)
+        }
+    }
+
+    /// Return the name of this `RecFuncDecl`.
+    ///
+    /// Strings will return the `Symbol`.  Ints will have a `"k!"` prepended to
+    /// the `Symbol`.
+    pub fn name(&self) -> String {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_ctx = self.ctx.z3_ctx;
+            let symbol = Z3_get_decl_name(z3_ctx, self.z3_func_decl);
+            match Z3_get_symbol_kind(z3_ctx, symbol) {
+                SymbolKind::String => CStr::from_ptr(Z3_get_symbol_string(z3_ctx, symbol))
+                    .to_string_lossy()
+                    .into_owned(),
+                SymbolKind::Int => format!("k!{}", Z3_get_symbol_int(z3_ctx, symbol)),
+            }
+        }
+    }
+}
+
+impl<'ctx> fmt::Display for RecFuncDecl<'ctx> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        let p = unsafe { Z3_func_decl_to_string(self.ctx.z3_ctx, self.z3_func_decl) };
+        if p.is_null() {
+            return Result::Err(fmt::Error);
+        }
+        match unsafe { CStr::from_ptr(p) }.to_str() {
+            Ok(s) => write!(f, "{}", s),
+            Err(_) => Result::Err(fmt::Error),
+        }
+    }
+}
+
+impl<'ctx> fmt::Debug for RecFuncDecl<'ctx> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        <Self as fmt::Display>::fmt(self, f)
+    }
+}
+
+impl<'ctx> Drop for RecFuncDecl<'ctx> {
+    fn drop(&mut self) {
+        unsafe {
+            Z3_dec_ref(
```

Comment:
This might need a mutex lock -- not sure that z3 reference counters are atomic.

Reply:
I'm not sure either, I copied that impl from FuncDecl https://github.com/prove-rs/z3.rs/blob/8b77fc554def507162a3860bea659247f2939859/z3/src/func_decl.rs#L128-L134 - should I aquire Z3_MUTEX?

Reply:
Ah ok, that's fine then. We can continue that thread in #143